### ISSUE TITLE
Folded definitions in swagger.json are not properly parsed

### ISSUE THREAD

**Issue by Somebi** [go-swagger/go-swagger#1289]
This is not working:
```
        "someIds": {
            "type": "array",
            "example": [
                413,
                231,
                777
            ],
            "default": [
                413,
                231,
                777
            ],
            "items": {
                "type": "number"
            }
        },
        "getSomeIds": {
            "type": "object",
            "properties": {
                "someIds": {
                    "$ref": "#/definitions/someIds",
                    "type": "array"
                }
            }
        },
```

This is working:
```
 "getSomeIds": {
            "type": "object",
            "properties": {
                "someIds": {
                    "type": "array",
                    "example": [
                        413,
                        231,
                        777
                    ],
                    "default": [
                        413,
                        231,
                        777
                    ],
                    "items": {
                        "type": "number"
                    }
                }
            }
        },
```

Inner/folded definitions should be parsed properly.

**Comment by gregmarr** [comment]
I'm not sure if this is the cause of your problem, but you can't have anything else with the `"$ref"`.

**Comment by Somebi** [comment]
What do you mean?

**Comment by beckqq** [comment]
It seems it's allowed by spec, probably json spec parser doesn't follow spec rules as intended.

**Comment by beckqq** [comment]
It seems it's allowed by spec, probably json spec parser doesn't follow spec rules as intended.

**Comment by Somebi** [comment]
It seems it's allowed by spec, probably json spec parser doesn't follow spec rules as intended.

**Comment by gregmarr** [comment]
https://github.com/swagger-api/swagger-editor/issues/902
>When you are using a $ref, it should'n have a sibling, otherwise you'll get a warning (it's weird that I didn't!). That's because when there is a $ref, everything (including $ref itself) will be replaced with the result of what that $ref is pointing to. So if you put siblings next to a $ref, they will not appear because resolver replaces it.

**Comment by gregmarr** [comment]
I expect that this isn't actually related to your problem, but it would be good to eliminate it to be sure.

**Comment by casualjim** [comment]
can you define "not working" a little bit clearer

**Comment by Somebi** [comment]
It's not generating server code with it.

**Comment by Somebi** [comment]
You can try to fold refs like i have posted and try yourself.

**Comment by fredbi** [comment]
Reproduced the different cases.
- $ref is out of cause (in this limited context at least)
- with correct $ref setting (e.g. without siblings in a schema, as required by Swagger 2.0 Spec - WHICH IS NOT json schema), spec validates and generates a server that builds
- issue can be pin-pointed to forbidden mix $ref + sibling, which is correctly detected by `swagger validate`, although the returned error message does not state clearly the issue (complains about missing "items")

Will add these fixtures in cmd/swagger/commands/validate_test.go for the record, then close the issue, if @somebi has no objection.

Fred

**Comment by Somebi** [comment]
issue can be pin-pointed to forbidden mix $ref + sibling, which is correctly detected by swagger validate, although the returned error message does not state clearly the issue (complains about missing "items")

Yes it was complaining about missing items if i'm not wrong and my memory is still good. :D

**Comment by fredbi** [comment]
This the actual error. 
Spec validation error messages are really poor and a great cause for misunderstanding and waste of time overall. An issue is now dedicated to this. 
At some point, there was also an idea for an enhancement to produce the line number in the spec file, but it seems quite a bit complicated.

I won't close this one, but requalify it as a request for enhancement with error messages, which seems fair enough.

**Comment by fredbi** [comment]
Will work next week on a PR to improve this, hopefully together with some other `swagger validate` shortcomings.

**Comment by fredbi** [comment]
@Somebi I improved somewhat the error signaling, but this fix does not (yet) identify the actual cause ($ref sibling). It is a bit complicated in the validation package to get it right.
Further, $ref siblings actually should generate a warning (they are not strictly forbidden, but the resulting behavior is most likely unwanted).

In your case, I could however croak some more valuable information about the problem, such as mising items in array, which should help in spec problem pin-pointing.

Bottom line: the ticket remains open, but we have somewhat improved.
